Use metabase-lib v2 to manage query row limit

kulyk opened the pull request:
Makes the `QuestionRowCount` use metabase-lib v2 (written in CLJC) to manage query limit.

> **Warning**
> A few unit-tests are failing because of a query conversion issue that will be fixed by #29520

![CleanShot 2023-03-22 at 18 47 31@2x](https://user-images.githubusercontent.com/17258145/227014291-6b50c2d6-af44-423a-a8ab-d746c2a96993.png)

The end goal is to encapsulate query manipulation complexity in a shared CLJC library to be used on both BE and FE. We want to keep the FE  isolated from query implementation details as much as possible. For instance, MLv2's query object would be opaque for the FE, and it'd be really hard to do something about it without using MLv2 helpers (see [here](https://github.com/metabase/metabase/pull/29454#discussion_r1145253129)).

The first steps in the FE migration are to make well-isolated components use MLv2 (I have row limit, summarization sidebar, filter modal, and maybe drills in mind. Notebook editor is more complex, and I think it'd make sense to approach it once we successfully migrate at least aggregations and filters). 

Most QB components work like that:

1. Receive props like `question` / `query` and `onChangeQuestion` /  `onChangeQuery`
2. Generate a new question/query using their methods (e.g. `const nextQuery = query.addFilter(filter)`
3. Call `onChangeQuestion` /  `onChangeQuery` with a new question/query instance
4. Most of the time it'd result in dispatching [the `updateQuestion` redux action](https://github.com/metabase/metabase/blob/02426a0cc9f95e8949bdb0d9aa7dc23f1dbd3dce/frontend/src/metabase/query_builder/actions/core/updateQuestion.ts#L110) that'd update the question in QB's redux store
5. The entire QB react would react to a question/query change

My idea is to make these components do something like that instead:

1. Retrieve a question/query from Redux to address the prop-drilling issue (see [this comment](https://github.com/metabase/metabase/pull/29454#discussion_r1145257379)). If possible, it's preferred to only return the data needed by a component instead of the whole question/query object (e.g. only the row limit for the limit component)
2. Add a redux action(s) that would mutate the query and make the component dispatch that action instead
3. The action would essentially take a question instance from QB redux state, retrieve an MLv2 query from it, update it, get a new MLv2 query, convert it to an MLv1 query, and dispatch the `updateQuestion` question

### Questions

1. Is `metabase-lib/v2` a good place to keep the TypeScript wrapper library?
7. How do we feel about adding thin helpers to the wrapper library ([see `hasLimit` in `metabase-lib/v2/limit`](https://github.com/metabase/metabase/pull/29454#discussion_r1145250155))
8. What do you think about pulling out query manipulation logic out of QB components to redux actions?
9. Any other kind of feedback is absolutely welcome

<!-- Reviewable:start -->
- - -
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/metabase/metabase/29454)
<!-- Reviewable:end -->

kulyk committed:
Add TypeScript wrapper for limit helpers

kulyk committed:
Use v2 limit in `StructuredQuery`

kulyk committed:
Add redux action for managing limit in QB

kulyk committed:
Switch `QuestionRowCount` to use ML v2

kulyk commented on frontend/src/metabase-lib/v2/limit.ts:
```diff
@@ -0,0 +1,15 @@
+import * as MetabaseLibLimit from "cljs/metabase.lib.limit";
+import type { Query, Limit } from "./types";
+
+export function currentLimit(query: Query): Limit {
+  return MetabaseLibLimit.current_limit(query);
+}
+
+export function limit(query: Query, limit: Limit): Query {
+  return MetabaseLibLimit.limit(query, limit);
+}
+
+export function hasLimit(query: Query) {
+  const limit = currentLimit(query);
+  return typeof limit === "number" && limit > 0;
+}
```
Not sure how cool we are about adding thin helpers in a wrapping library. Should this better be done in CLJS? 🤔

snoe replied:
yeah, I think it makes sense to add to cljs.

bshepherdson replied:
I'm not philosophically opposed to a bit of TS logic, but in this case the function is useful in Clojure too, so it may as well live there.

kulyk commented on frontend/src/metabase-lib/v2/types.ts:
```diff
@@ -0,0 +1,3 @@
+export type Query = unknown;
```
In v2, the query object should be totally opaque for the FE, i.e. it'd be _really_ hard to work with directly without using lib functions. I'm not sure how to properly type to indicate it's not meant to be accessed directly but to also make sure we're passing around expected things

FWIW, here's how a v2 query looks like on the FE right now:

<img width="559" alt="CleanShot 2023-03-22 at 18 33 59@2x" src="https://user-images.githubusercontent.com/17258145/227003810-9e52f469-701d-4bd2-9ab9-975b7933d860.png">

metamben replied:
> make sure we're passing around expected things

Are these expected things context dependent? I.e., would it be intractable to encode the requirements into opaque types?

kulyk replied:
@metamben I don't think typing e.g. a query object returned from MLv2 would make sense. I just want TypeScript to make sure that when I call `limit(query)`, it'd expect `query` to be what `question._getMLv2Query` returns

bshepherdson replied:
Yeah, there's a trick to do that in TS. It makes distinct, "unforgeable" types that can only be created in the same file. 

We don't need to do it in this PR, but we can easily make it so all the TS wrappers in `frontend/src/metabase-lib/` that call CLJC things return these opaque types. then TS will check that eg. what you pass to `Lib.limit(q, 10)` must be the opaque `Query` (or whatever) type returned by other MLv2 functions. (Of course you can always `as unknown as Query` in TS but you'll get the bugs you deserve in that case.)

bshepherdson replied:
Reference: https://calebmer.com/tangents/opaque-types-in-typescript.html
Example in a pending PR: https://github.com/metabase/metabase/blob/e0a93a89068be2ba688dbeadb44505d15056ab7f/frontend/src/metabase-lib/Question.ts#L114

ranquild replied:
Agree with @bshepherdson , we could use opaque TS types for this. `unknown` has downsides as anything satisfies `unknown`

kulyk replied:
[`Query` is now opaque](https://github.com/metabase/metabase/blob/c99ca5ac4fcd847c8bc8f77225dbd9dbba367211/frontend/src/metabase-lib/types.ts#L6). Passing another opaque type instead of a `Query` would still fail. Only two ways to bypass it are `@ts-ignore` or `as Query` which should be normally caught during the PR review

kulyk commented on frontend/src/metabase/query_builder/actions/query-updates.ts:
```diff
@@ -0,0 +1,20 @@
+import { getQuestion } from "../selectors";
+import { updateQuestion } from "./core";
+
+export const setLimit =
+  (limit: Limit) => (dispatch: Dispatch, getState: GetState) => {
+    const state = getState();
+    const question = getQuestion(state);
+    if (!question) {
+      return;
+    }
+    const query = question._getMLv2Query();
+    const nextV2Query = MetabaseLib.limit(query, limit);
+    const nextV1Query = MetabaseLib.toV1Query(nextV2Query);
+    dispatch(updateQuestion(nextV1Query, { run: true }));
+  };
```
Note: `MetabaseLib.toV1Query` doesn't exist yet. I just wanted to show the idea that we'd need to convert queries between v1 and v2 formats until we fully migrate Metabase to metabase-lib v2

kulyk commented on frontend/src/metabase/query_builder/components/notebook/lib/steps.js:
```diff
@@ -93,7 +93,7 @@ const STEPS = [
     active: query => query.hasLimit(),
     revert: query => query.clearLimit(),
-    clean: query => query.cleanLimit(),
+    clean: query => query,
```
This PR removes the `cleanLimit` method because it didn't do anything

kulyk commented on frontend/src/metabase/query_builder/components/view/QuestionRowCount.jsx:
I think QB migration to MLv2 could be a good opportunity to at least partially address QB prop drilling. As for now, [the root `QueryBuilder` component ](https://github.com/metabase/metabase/blob/master/frontend/src/metabase/query_builder/containers/QueryBuilder.jsx) maps every QB action and selector to it and passes all the props down. Would appreciate some feedback on that from FE folks

snoe commented on frontend/src/metabase-lib/v2/limit.ts:
```diff
@@ -0,0 +1,15 @@
+import * as MetabaseLibLimit from "cljs/metabase.lib.limit";
```
I imagined we'd only import from `metabase.lib.core` as `MetabaseLib`

kulyk replied:
Fixed to use `metabase.lib.core` 👍

kulyk commented on frontend/src/metabase/query_builder/components/view/QuestionRowCount.jsx:
We'd most likely want to add test coverage to files like that first before migrating to MLv2

snoe commented on frontend/src/metabase-lib/v2/limit.ts:
I suspect, that we want to start using `stage-number` (optional arg after `query`) earlier than later.

kulyk commented on frontend/src/metabase/query_builder/actions/query-updates.ts:
```diff
@@ -0,0 +1,20 @@
+import type { Dispatch, GetState } from "metabase-types/store";
+
+import * as MetabaseLib from "metabase-lib/v2";
```
FE folks, do you have any naming preferences here? `MetabaseLib` feels a bit too massive; how about just `Lib` or `ML`? Other ideas are welcome

kulyk committed:
Add utils to convert v2 queries into v1

kulyk committed:
Use `toLegacyQuery`

bshepherdson commented on frontend/src/metabase/query_builder/actions/query-updates.ts:
```diff
@@ -0,0 +1,21 @@
+    if (!question) {
+      return;
+    }
+    const query = question._getMLv2Query();
+    const nextQuery = MetabaseLib.limit(query, limit);
+    const nextLegacyQuery = MetabaseLib.toLegacyQuery(nextQuery);
+    const nextQuestion = question.setDatasetQuery(nextLegacyQuery);
+    dispatch(updateQuestion(nextQuestion, { run: true }));
```
This will feel a lot better when we no longer need the `Question` wrapper and this becomes just
```ts
const query = getQuery(state);
if (query) {
  dispatch(updateQuery(MetabaseLib.limit(query, limit), { run: true }));
}
```

kulyk replied:
Yeah, let's see how it looks once we add a few more actions here, maybe we'd generalize them for some time

bshepherdson commented on frontend/src/metabase/query_builder/components/view/QuestionRowCount.jsx:
```diff
@@ -70,11 +101,11 @@ const QuestionRowCount = ({
-              if (limit > 0) {
-                onQueryChange(query.updateLimit(limit));
+            onChangeLimit={nextLimit => {
+              if (nextLimit > 0) {
+                onChangeLimit(nextLimit);
               } else {
-                onQueryChange(query.clearLimit());
+                onChangeLimit(null);
               }
```
This is one of my pet peeves 😜  how about
```ts
onChangeLimit{nextLimit => {
  onChangeLimit(nextLimit > 0 ? nextLimit : null);
}}
```
?

kulyk replied:
Makes total sense, fixed!

bshepherdson reviewed (commented):
Broadly looks good to me! (Both the `metabase-lib/` interactions, and how it plays out in the component and Redux actions.)

The only thing that's a must-change in my book is the opaque type thing. We can do a lot better than aliases for `unknown`!

kulyk committed:
Fix CLJS limit functions unavailable in Jest

kulyk committed:
Fix tests

kulyk committed:
Style tweaks

kulyk committed:
Fix type error

kulyk committed:
Use opaque type for ML `Query`

kulyk committed:
Use `Lib` instead of `MetabaseLib`

kulyk committed:
Drop v2 folder

kulyk commented:
> > Is metabase-lib/v2 a good place to keep the TypeScript wrapper library?
>
> I'd rather avoid this. In due course most of what's currently in metabase-lib/ will be deleted, replaced or rewritten. After that transition, the v2 distinction is noise. Let's just put things in sensibly named files under metabase-lib/ and clean up as we go.

@bshepherdson sounds good to me. I removed the `v2` library, and all the new files now live under `metabase-lib`. Also added an `index` file there which should only export v2 stuff. I think we'd expect it to be used something like that:

```js
import * as Lib from "metabase-lib"
import Question from "metabase-lib/Question"
```

bshepherdson commented on frontend/src/metabase-lib/limit.ts:
```diff
@@ -0,0 +1,18 @@
+import {
+  current_limit as _current_limit,
+  limit as _limit,
+} from "cljs/metabase.lib.core";
```
This isn't bad here, but this feels like it's going to get really clunky in larger files. Is there a reason (style conventions?) not to do `import * as ML` or similar?

bshepherdson replied:
Does it still? I thought that was an early issue because the `lib.limit/*` functions weren't being re-exported in `src/metabase/lib/core.cljc` (they are now).

kulyk replied:
@bshepherdson just tried it once more after rebasing on the latest `master`, and I'm still getting a `ML.limit is not a function` kind of error when I'm doing `import * as ML from "metabase.lib.core"

bshepherdson commented on frontend/src/metabase-lib/queries/StructuredQuery.ts:
```diff
@@ -49,11 +48,16 @@ import Dimension, {
 } from "metabase-lib/Dimension";
 import DimensionOptions from "metabase-lib/DimensionOptions";
+
+import * as MLv2 from "..";
```
`as ML`?

bshepherdson commented on frontend/src/metabase/query_builder/actions/query-updates.ts:
```diff
@@ -0,0 +1,21 @@
+export const setLimit =
+  (limit: Limit) => (dispatch: Dispatch, getState: GetState) => {
+    const state = getState();
+    const question = getQuestion(state);
+    if (!question) {
+      return;
+    }
+    const query = question._getMLv2Query();
```
This feels like it should be its own selector, but that doesn't need to be now.

kulyk replied:
Yeah, agree. I'd like to see how it looks with a couple more things ported before trying to generalize/abstract this

bshepherdson reviewed (approved):
> Reviewed 18 of 18 files at r1, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/metabase/metabase/29454)* status: all files reviewed, 3 unresolved discussions (waiting on @alxnddr, @iethree, @kulyk, @metamben, @ranquild, and @snoe)

*[a discussion](https://reviewable.io/reviews/metabase/metabase/29454#gh-1146240884:-NRdq0i6-aMXa3a6m7mN:bvbjvmh) (no related file):*
<details><summary><i>Previously, iethree (Ryan Laurie) wrote…</i></summary><blockquote>

`M9L` 😂 
</blockquote></details>

Seems like this is settling on `import * as Lib from "metabase-lib";` which seems good to me.

___


<!-- Sent from Reviewable.io -->

kulyk committed:
Fix MBQL in unit tests

kulyk committed:
Don't import individual functions from MLv2

kulyk committed:
Export `limit` CLJC functions

kulyk committed:
Tweak CLJS imports

bshepherdson reviewed (approved):
> Reviewed 5 of 5 files at r2, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/metabase/metabase/29454)* status: all files reviewed, 3 unresolved discussions (waiting on @alxnddr, @iethree, @kulyk, @metamben, @ranquild, and @snoe)

*[a discussion](https://reviewable.io/reviews/metabase/metabase/29454#gh-1149324863:-NRiBNmiEtIS1jxSdzyI:b-2dndgp) (no related file):*
<details><summary><i>Previously, kulyk (Anton Kulyk) wrote…</i></summary><blockquote>

@bshepherdson just tried it once more after rebasing on the latest `master`, and I'm still getting a `ML.limit is not a function` kind of error when I'm doing `import * as ML from "metabase.lib.core"
</blockquote></details>

I see that you've fixed it as we discussed. <img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> 

___


<!-- Sent from Reviewable.io -->
